src/js/addListeners.js

```diff
@@ -0,0 +1,129 @@
+'use strict';
+
+var $ = require('jquery');
+var validate = require('./validate');
+var getErrorElement = require('./helpers/getErrorElement');
+
+$(document).on('keyup click change blur', '[data-validations]', onChange);
+
+function onChange(e) {
+  var $input = $(this);
+  var $error = getErrorElement($input);
+
+  var invalidValue = validate.element(this);
+  var errorNeeded = invalidValue && $input.hasClass('is-dirty');
+  var inputInitiallyEmpty =  !$input.hasClass('is-filled');
+
+  // who knows what this is for
+  if (!$input.attr('data-validations') && $input.val()) {
+    $input.addClass('is-dirty is-valid');
+  }
+
+  if (inputInitiallyEmpty) {
+    return;
+  }
+
+  if (e.type === 'keyup' && !$input.hasClass('is-filled')) {
+    return;
+  }
+
+  if (invalidValue) {
+    $error.text(invalidValue);
+  } else {
+    $error.text('');
+    showError(false);
+  }
+
+  if (errorNeeded) {
+    showError(true);
+  }
+
+  // Do not use getParentForm()
+  validate.form($input.parents('form').get(0));
+
+  function showError(showError) {
+    if (
+      $input.data('errorasopacity')
+      && !$error.css('position') === 'relative'
+    ) {
+      $error
+        .css('visibility', showError ? 'visible' : 'hidden')
+        .fadeTo(showError ? 1 : 0);
+    } else {
+      $error[showError ? 'show' : 'hide']()
```

Comment:
Can't you use `.toggle()` with a boolean attribute here? http://api.jquery.com/toggle/#toggle-display